subsys/net/ip/ipv4_autoconf.c

```diff
@@ -0,0 +1,263 @@
+/** @file
+ * @brief IPv4 autoconf related functions
+ */
+
+/*
+ * Copyright (c) 2017 Matthias Boesl
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_NET_DEBUG_IPV4_AUTOCONF)
+#define SYS_LOG_DOMAIN "net/ipv4ll"
+#define NET_LOG_ENABLED 1
+#endif
+
+#include "net_private.h"
+#include <errno.h>
+#include <net/arp.h>
+#include <net/net_pkt.h>
+#include <net/net_core.h>
+#include <net/net_if.h>
+
+#include <net/ipv4_autoconf.h>
+
+static void send_probe(struct net_if *iface);
+static void send_announcement(struct net_if *iface);
+
+static inline struct net_pkt *prepare_arp(struct net_if *iface)
+{
+	struct net_pkt *pkt;
+	struct net_buf *frag;
+	struct net_arp_hdr *hdr;
+	struct net_eth_hdr *eth;
+
+	pkt = net_pkt_get_reserve_tx(sizeof(struct net_eth_hdr), K_FOREVER);
+	if (!pkt) {
+		goto fail;
+	}
+
+	frag = net_pkt_get_frag(pkt, K_FOREVER);
+	if (!frag) {
+		goto fail;
+	}
+
+	net_pkt_frag_add(pkt, frag);
+	net_pkt_set_iface(pkt, iface);
+	net_pkt_set_family(pkt, AF_INET);
+	net_pkt_set_ll_reserve(pkt, sizeof(struct net_eth_hdr));
+
+	eth = NET_ETH_HDR(pkt);
+	hdr = NET_ARP_HDR(pkt);
+	eth->type = htons(NET_ETH_PTYPE_ARP);
+
+	/* ethernet header */
+	memcpy(&eth->src.addr, net_if_get_link_addr(iface)->addr,
+	       sizeof(struct net_eth_addr));
+	memset(&eth->dst.addr, 0xFF, sizeof(struct net_eth_addr));
+
+	hdr->hwtype = htons(NET_ARP_HTYPE_ETH);
+	hdr->protocol = htons(NET_ETH_PTYPE_IP);
+	hdr->hwlen = sizeof(struct net_eth_addr);
+	hdr->protolen = sizeof(struct in_addr);
+	hdr->opcode = htons(NET_ARP_REQUEST);
+
+	memcpy(&hdr->src_hwaddr.addr, &eth->src.addr,
+	       sizeof(struct net_eth_addr));
+	memset(&hdr->dst_hwaddr.addr, 0x00, sizeof(struct net_eth_addr));
+
+	net_ipaddr_copy(&hdr->dst_ipaddr, &iface->config.ipv4auto.current_ip);
+	net_ipaddr_copy(&hdr->src_ipaddr, &iface->config.ipv4auto.requested_ip);
+
+	net_buf_add(frag, sizeof(struct net_arp_hdr));
+
+	return pkt;
+
+fail:
+	net_pkt_unref(pkt);
+	return NULL;
+}
+
+enum net_verdict net_ipv4_autoconf_input(struct net_if *iface,
+					 struct net_pkt *pkt)
+{
+	struct net_arp_hdr *arp_hdr;
+
+	if (net_pkt_get_len(pkt) < (sizeof(struct net_arp_hdr) -
+				    net_pkt_ll_reserve(pkt))) {
+		NET_DBG("Invalid ARP header (len %zu, min %zu bytes)",
+			net_pkt_get_len(pkt),
+			sizeof(struct net_arp_hdr) -
+			net_pkt_ll_reserve(pkt));
+		return NET_DROP;
+	}
+
+	arp_hdr = NET_ARP_HDR(pkt);
+
+	if (!net_ipv4_addr_cmp(&arp_hdr->dst_ipaddr,
+			       &iface->config.ipv4auto.requested_ip)) {
+		/* No conflict */
+		return NET_DROP;
+	}
+
+#if NET_DEBUG > 0
+	do {
+		NET_DBG("conflict detected from %s for %s",
+			net_sprint_ll_addr((uint8_t *)&arp_hdr->src_hwaddr,
+					   arp_hdr->hwlen),
+			net_sprint_ipv4_addr(&arp_hdr->dst_ipaddr));
+	} while (0);
+#endif
+
+	iface->config.ipv4auto.conflict_cnt++;
+
+	switch (iface->config.ipv4auto.state) {
```

Comment:
Instead of iface->config, do net_if_get_config(iface). Do the change to rest of the file too.